nuxeo-core/nuxeo-core-storage-dbs/src/main/java/org/nuxeo/ecm/core/storage/dbs/DBSCachingRepository.java

```diff
@@ -0,0 +1,399 @@
+        }
+        try {
+            return klass.newInstance();
+        } catch (ReflectiveOperationException e) {
+            throw new NuxeoException(e);
+        }
+
+    }
+
+    public void begin() {
+        repository.begin();
+        processReceivedInvalidations();
+    }
+
+    @Override
+    public void commit() {
+        repository.commit();
+        sendInvalidationsToOther();
+        processReceivedInvalidations();
+    }
+
+    @Override
+    public void rollback() {
+        repository.rollback();
+    }
+
+    @Override
+    public void shutdown() {
+        repository.shutdown();
+        cache.invalidateAll();
+        childCache.invalidateAll();
+        // Send invalidations
+        if (clusterInvalidator != null) {
+            clusterInvalidator.sendInvalidations(new DBSInvalidations(true));
+        }
+
+    }
+
+    @Override
+    public State readState(String id) {
+        State state = cache.getIfPresent(id);
+        if (state == null) {
+            state = repository.readState(id);
+            if (state != null) {
+                putInCache(state);
+            }
+        }
+        return state;
+    }
+
+    @Override
+    public List<State> readStates(List<String> ids) {
+        ImmutableMap<String, State> statesMap = cache.getAllPresent(ids);
+        List<String> idsToRetrieve = new ArrayList<>(ids);
+        idsToRetrieve.removeAll(statesMap.keySet());
+        // Read missing states from repository
+        List<State> states = repository.readStates(idsToRetrieve);
+        // Cache them
+        states.forEach(this::putInCache);
+        // Add previous cached one
+        states.addAll(statesMap.values());
+        // Sort them
+        states.sort(Comparator.comparing(state -> state.get(KEY_ID).toString(), Ordering.explicit(ids)));
+        return states;
+    }
+
+    @Override
+    public void createState(State state) {
+        repository.createState(state);
+        // TODO check if it's relevant
+        putInCache(state);
+    }
+
+    @Override
+    public void createStates(List<State> states) {
+        repository.createStates(states);
+        states.forEach(this::putInCache);
+    }
+
+    @Override
+    public void updateState(String id, StateDiff diff) {
+        repository.updateState(id, diff);
+        invalidate(id);
+    }
+
+    @Override
+    public void deleteStates(Set<String> ids) {
+        repository.deleteStates(ids);
+        invalidateAll(ids);
+    }
+
+    @Override
+    public State readChildState(String parentId, String name, Set<String> ignored) {
+        processReceivedInvalidations();
+
+        String childCacheKey = computeChildCacheKey(parentId, name);
+        String stateId = childCache.getIfPresent(childCacheKey);
+        if (stateId != null) {
+            State state = cache.getIfPresent(stateId);
+            if (state != null && parentId.equals(state.get(KEY_PARENT_ID)) && name.equals(state.get(KEY_NAME))) {
```

Comment:
Please add a comment about why we do this: the cache could be invalid (due to a move of the doc to a different parent or name) and we have no invalidations for the childCache.